src/transaction.ts

```diff
@@ -0,0 +1,359 @@
+   *   s: '0x5bd428537f05f9830e93792f90ea6a3e2d1ee84952dd96edbae9f658f831ab13'
+   * };
+   * const tx = new Transaction(txData);
+   * ```
+   */
+  constructor(
+    data: Buffer | PrefixedHexString | BufferLike[] | TxData = {},
+    opts: TransactionOptions = {},
+  ) {
+    // instantiate Common class instance based on passed options
+    if (opts.common) {
+      if (opts.chain) {
+        throw new Error('Instantiation with both opts.common and opts.chain parameter not allowed!')
+      }
+      this._common = opts.common
+    } else {
+      const chain = opts.chain ? opts.chain : 'mainnet'
+      const hardfork = opts.hardfork ? opts.hardfork : 'byzantium'
+      this._common = new Common(chain, hardfork)
+    }
+
+    // Define Properties
+    const fields = [
+      {
+        name: 'nonce',
+        length: 32,
+        allowLess: true,
+        default: new Buffer([]),
+      },
+      {
+        name: 'gasPrice',
+        length: 32,
+        allowLess: true,
+        default: new Buffer([]),
+      },
+      {
+        name: 'gasLimit',
+        alias: 'gas',
+        length: 32,
+        allowLess: true,
+        default: new Buffer([]),
+      },
+      {
+        name: 'to',
+        allowZero: true,
+        length: 20,
+        default: new Buffer([]),
+      },
+      {
+        name: 'value',
+        length: 32,
+        allowLess: true,
+        default: new Buffer([]),
+      },
+      {
+        name: 'data',
+        alias: 'input',
+        allowZero: true,
+        default: new Buffer([]),
+      },
+      {
+        name: 'v',
+        allowZero: true,
+        default: new Buffer([opts.chain || opts.common ? this._common.chainId() : 0x1c]),
+      },
+      {
+        name: 'r',
+        length: 32,
+        allowZero: true,
+        allowLess: true,
+        default: new Buffer([]),
+      },
+      {
+        name: 's',
+        length: 32,
+        allowZero: true,
+        allowLess: true,
+        default: new Buffer([]),
+      },
+    ]
+
+    /**
+     * Returns the transaction in JSON format
+     * @method toJSON
+     * @return {Array | String}
+     * @memberof Transaction
+     * @name toJSON
+     * @see {@link https://github.com/ethereumjs/ethereumjs-util/blob/master/docs/index.md#defineproperties|ethereumjs-util}
+     */
+
+    // attached serialize
+    defineProperties(this, fields, data)
+
+    /**
+     * @property {Buffer} from (read only) sender address of this transaction, mathematically derived from other parameters.
+     * @name from
+     * @memberof Transaction
+     */
+    Object.defineProperty(this, 'from', {
+      enumerable: true,
+      configurable: true,
+      get: this.getSenderAddress.bind(this),
+    })
+
+    // calculate chainId from signature
+    const sigV = bufferToInt(this.v)
+    let chainId = Math.floor((sigV - 35) / 2)
+    if (chainId < 0) chainId = 0
+
+    // set chainId
+    if (opts.chain || opts.common) {
```

Comment:
Sorry this is not directly related to your PR, but could you maybe add a short note to the tsdoc that explains the order of priority for chainId? if I'm not mistaken its 1. `opts.chain || opts.common.chainId()`, 2. chainId from signature and lastly `data.chainId` 4. defaults to 0! Not sure why so complex.

Reply:
I'll document it if we decide to keep it. I agree that it's too complex.

Reply:
//cc @danjm

Reply:
See related comment from @danjm here: https://github.com/ethereumjs/ethereumjs-tx/pull/131#issuecomment-478304415

Reply:
I understand this has been to avoid breaking changes and for the convenience of the user. I do think however that it makes it very easy for the user to make mistakes given that a single value can be set from 4 sources in an arbitrary order. What I propose:

Make `common` a mandatory option, remove `opts.chain`, `data.chainId` and throw if `common.chainId()` doesn't match the one from signature (if present). Making `common` mandatory can also allow us to deprecate `chain`/`hardfork` and allow users to migrate.

That said, I don't feel very strongly and this was simply my (relatively uninformed) opinion.

Reply:
Yes, we should definitely simplify here before the release. On first thought I would go along with your proposal.

We should put some additional reflection on the question if there are disadvantages if we differ here from the pattern regarding the ``Common`` respectively ``chain``/``hardfork`` parameter passing compared to the other libraries. But probably it's ok (?).

Reply:
@alcuadrado what do you think? We can address this in a separate PR though but whatever eases the process.

Reply:
I'm not sure if making common mandatory is the best, as users that aren't familiar with it will find it confusing.

One option is to throw whenever there are incompatible sets of options. E.g. if you pass it `common` and `chain`, throw despite their values.

I opened https://github.com/ethereumjs/ethereumjs-tx/issues/149 to continue this discussion.

Reply:
> if you pass it common and chain, throw despite their values.

Just a note that this is already done here: https://github.com/ethereumjs/ethereumjs-tx/blob/master/src/transaction.ts#L70

Reply:
Ok, then let's move on here with keeping ``Common`` and alternatively ``chain``/``hardfork`` passing "as is" together with the throw mechanism on incompatible options @danjm mentioned. Otherwise we'll implement what Sina proposed above:

> Make `common` a mandatory option, remove `opts.chain`, `data.chainId` and throw if `common.chainId()` doesn't match the one from signature (if present). Making `common` mandatory can also allow us to deprecate `chain`/`hardfork` and allow users to migrate.

And we then access ``chainId`` solely through the passed or generated ``Common`` instance.

Does this make sense?

What is with this ``0`` default value. Is this necessary respectively does it make sense?

Reply:
@alcuadrado Can you prepare a PR on this once we have an agreement?